src/Language/PureScript/CoreFn/Desugar.hs

```diff
@@ -132,9 +132,14 @@ moduleToCoreFn env (A.Module _ coms mn decls (Just exps)) =
   altToCoreFn :: Maybe SourceSpan -> A.CaseAlternative -> CaseAlternative Ann
   altToCoreFn ss (A.CaseAlternative bs vs) = CaseAlternative (map (binderToCoreFn ss []) bs) (go vs)
     where
-    go :: Either [(A.Guard, A.Expr)] A.Expr -> Either [(Guard Ann, Expr Ann)] (Expr Ann)
-    go (Left ges) = Left $ map (exprToCoreFn ss [] Nothing *** exprToCoreFn ss [] Nothing) ges
-    go (Right e) = Right (exprToCoreFn ss [] Nothing e)
+    go :: [A.GuardedExpr] -> Either [(Guard Ann, Expr Ann)] (Expr Ann)
+    go [A.MkUnguarded e]
+      = Right (exprToCoreFn ss [] Nothing e)
+    go gs
+      = Left [ (exprToCoreFn ss [] Nothing cond, exprToCoreFn ss [] Nothing e)
+             | A.GuardedExpr g e <- gs
+             , let [A.ConditionGuard cond] = g -- TODO: check invariant
```

Comment:
I think I meant to give some meaningful error when pattern guard is encountered.